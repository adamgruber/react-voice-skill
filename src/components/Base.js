class BaseComponent {
  constructor(root, props = {}) {
    console.log('BaseComponent constructor');
    // console.log(root);
    // console.log(props);
    // console.log('\n');
    this.children = [];
    this.root = root;
    this.props = props;
    this.type = props.$$type;
  }

  appendChild(child) {
    this.validateChild(child);
    this.children.push(child);
  }

  removeChild(child) {
    console.log('BaseComponent removeChild');
    const index = this.children.indexOf(child);
    this.children.splice(index, 1);
  }

  getValidChildren() {}

  validateChild(child) {
    // Strings are always valid
    if (typeof child === 'string') return;

    const validChildren = this.getValidChildren();
    if (!validChildren) {
      throw new Error('Components must implement a getValidChildren method that returns an array of valid child types.');
    }

    if (!child.type) {
      throw new Error('Components must have a type.');
    }

    if (!validChildren.includes(child.type)) {
      throw new Error(`${child.type} is not a valid child of ${this.type}.`);
    }
  }
}

export default BaseComponent;
