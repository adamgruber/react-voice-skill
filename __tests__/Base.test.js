import Base from '../src/components/Base';

describe('Base Component', () => {
  let subject;
  beforeEach(() => {
    subject = new Base();
  });

  it('should have base properties', () => {
    expect(subject).toMatchSnapshot();
  });

  describe('appendChild', () => {
    beforeEach(() => {
      subject.validateChild = jest.fn();
      subject.appendChild();
    });

    it('should call validateChild', () => {
      expect(subject.validateChild).toHaveBeenCalled();
    });
  });

  describe('getValidChildren', () => {
    it('should return false', () => {
      expect(subject.getValidChildren()).toBe(undefined);
    });
  });

  describe('validateChild', () => {
    beforeEach(() => {
      subject.getValidChildren = jest.fn();
    });

    it('should throw if getValidChildren returns undefined', () => {
      expect(() => subject.validateChild({}))
        .toThrow('Components must implement a getValidChildren method that returns an array of valid child types.');
    });

    it('should throw if child.type is undefined ', () => {
      subject.getValidChildren.mockReturnValue([]);
      expect(() => subject.validateChild({}))
        .toThrow('Components must have a type.');
    });

    it('should throw if child.type is not valid', () => {
      subject.getValidChildren.mockReturnValue([]);
      expect(() => subject.validateChild({ type: 'TestComponent' }))
        .toThrow('TestComponent is not a valid child of undefined');
    });
  });
});
