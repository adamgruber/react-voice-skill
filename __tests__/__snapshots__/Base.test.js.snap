// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Base Component should have base properties 1`] = `
BaseComponent {
  "children": Array [],
  "props": Object {},
  "root": undefined,
  "type": undefined,
}
`;
